app-backend/tile/src/main/scala/LayerCache.scala

```diff
@@ -37,96 +40,92 @@ import scala.util._
   * things that require time to generate, usually a network fetch, use AsyncLoadingCache
   */
-object LayerCache extends Config {
+object LayerCache extends Config with LazyLogging {
   implicit val database = Database.DEFAULT
 
   val memcachedClient = KryoMemcachedClient.DEFAULT
 
   /** The caffeine cache to use for attribute stores */
-  val attributeStoreCache: ScaffeineCache[String, Future[S3AttributeStore]] =
+  private val attributeStoreCache: ScaffeineCache[String, AttributeStore] =
     Scaffeine()
       .recordStats()
       .expireAfterAccess(5.minutes)
       .maximumSize(500)
-      .build[String, Future[S3AttributeStore]]()
+      .build[String, AttributeStore]()
 
-  /**
-    * The primary means of interaction with this class: pass as key and
-    *  the costly option caching prevents duplication of.
-    */
-  def attributeStore(bucket: String, prefix: Option[String]): Future[S3AttributeStore] = {
-    val cacheKey = HeapBackedMemcachedClient.sanitizeKey(s"store-$bucket-$prefix")
-    attributeStoreCache.get(cacheKey, { cacheKey =>
-      prefix match {
-        case Some(prefixStr) => Future.successful(S3AttributeStore(bucket, prefixStr))
-        case None => Future.failed(new LayerIOError("Scene has no ingest location"))
-      }
-    })
-  }
+  private val histogramCache = HeapBackedMemcachedClient(memcachedClient)
+  private val tileCache = HeapBackedMemcachedClient(memcachedClient)
+  private val layerLocationCache = HeapBackedMemcachedClient(memcachedClient)
 
-  def attributeStore(prefix: Option[String]): Future[S3AttributeStore] =
-    attributeStore(defaultBucket, prefix)
+  def layerUri(layerId: UUID): Future[Option[String]] = {
+    layerLocationCache.caching(s"layer-uri-${layerId.toString}") { _ =>
+      Scenes.getScene(layerId).map(_.flatMap(_.ingestLocation))
+    }
+  }
 
-  val tileCache = HeapBackedMemcachedClient[Option[MultibandTile]](memcachedClient)
-  def maybeRenderExtent(id: UUID, zoom: Int, extent: Extent): Future[Option[MultibandTile]] =
-    tileCache.caching(s"rendered-extent-$id-$zoom-$extent") { implicit ec =>
-      for {
-        prefix <- prefixFromLayerId(id)
-        store <- attributeStore(defaultBucket, prefix)
-      } yield {
-        val reader = new S3ValueReader(store).reader[SpatialKey, MultibandTile](LayerId(id.toString, zoom))
-        val query = S3CollectionLayerReader(store)
-          .query[SpatialKey, MultibandTile, TileLayerMetadata[SpatialKey]](LayerId(id.toString, zoom))
-        val tile = query.where(Intersects(extent))
-          .result
-          .stitch
-          .crop(extent)
-          .tile
-        tile match {
-          case tile: MultibandTile => Some(tile)
-          case _ => None
-        }
+  def attributeStoreForLayer(layerId: UUID): Future[Option[AttributeStore]] = {
+    // we are not caching the attribute store based on layerId but based on the catalog location
+    OptionT(layerUri(layerId)).mapFilter { catalogUri =>
```

Comment:
This is good: http://typelevel.org/cats/datatypes/optiont.html

I like to think of monads like data types that encode a value (of some well understood underlying type) in some kind of context. `List`s (and `Option`s if you squint) do a loop-kind-of-thing, `Future`s encode a promise-sort-of-thing.

`flatMap` (haskell's `bind`), which is how for-comprehensions work under the covers, allow us to apply some function to the underlying value and make a decision based on that about how to manipulate (if at all) its wrapping, monadic context. A chain of flatmaps on `Option` which returns `None` at any point will fail fast: flatmapping over `None` will yield `None`.

This amazing feat - having turned an explicit and error-prone set of doings (code written to handle possibly existing values, abstracting over arity via loops, etc. etc.) into a description which sufficiently captures the semantics of said doings - falls apart when you start wrapping contexts. `Future[Option[T]]` is a promisey, possibly existing T, but we'd kind of like to deal with both modes of failure/success (not returning and being None/returning and being Some) without getting bogged down in nested for loops.